<a href="https://colab.research.google.com/github/Jeronimo-Giraldo/Analitica-de-negocios/blob/main/Reto_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**RETO #1-JERÓNIMO GIRALDO MARTÍNEZ**

0.Se procede con la carga de las librerías de trabajo.

In [ ]:
import numpy as np #Funciones numéricas para la transformación y análisis de los datos.
import pandas as pd #Conectar el sistema con las bases de datos.

1.Se procede con la carga de la base de datos.

In [ ]:
nxl="/content/1. BD2_CarCrash_Int.xlsx"
XDB=pd.read_excel(nxl,sheet_name=1)

#Se cargan las variables de interés.
XDB=XDB[['Weekday','ViolCat','ClearWeather','Month','CrashType','Highway','Daylight']]

##Separamos las variables por Entrada-Salida.
XD=XDB[['Weekday','ViolCat','ClearWeather','Month','CrashType','Daylight']] #Aquí están las variables de entrada.
yd=XDB[['Highway']] #Aquí está la variable de referencia.

2.Implementamos el Modelo.

In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB()
mnb.fit(XD,yd)

#Se obtiene la información:
ct=mnb.class_count_
print("**Cómo están separados los datos:**\n",ct)
print("**El total de los datos es:**\n",len(XD))
print("**Los datos porcentualmente están divididos en:**\n",(ct/len(XD))*100)

#Se obtiene la información por categoría:
u=mnb.theta_
print("**La media de los accidentes por categoría es:**\n",u)
std=mnb.var_  #Esto es Varianza
std=np.sqrt(std)
print("**La desviación estandar de los accidentes por categoría es:**\n",std)

#Se identifican los valores superiores e inferiores de los ingresos.
print("**El límite superior de los valores es:**\n",u+std)
print("**El límite inferior de los valores es:**\n",u-std)

#Para evaluar un solo caso:
XS=[[5,1,1,3,4,0]]
mnb.predict(XS)

**Cómo están separados los datos:**
 [1580.  501.]
**El total de los datos es:**
 2081
**Los datos porcentualmente están divididos en:**
 [75.92503604 24.07496396]
**La media de los accidentes por categoría es:**
 [[3.88417722 7.45696203 0.86265823 4.35189873 4.21772152 0.66265823]
 [3.9261477  4.1497006  0.79840319 6.69461078 3.27744511 0.62075848]]
**La desviación estandar de los accidentes por categoría es:**
 [[1.95062542 3.52136997 0.34420782 2.3433645  1.87371962 0.47280262]
 [2.01603723 2.16198902 0.40119266 3.44163641 1.05765448 0.48519831]]
**El límite superior de los valores es:**
 [[ 5.83480263 10.978332    1.20686605  6.69526323  6.09144114  1.13546084]
 [ 5.94218494  6.31168962  1.19959585 10.13624719  4.33509959  1.10595679]]
**El límite inferior de los valores es:**
 [[1.9335518  3.93559205 0.51845041 2.00853424 2.3440019  0.18985561]
 [1.91011047 1.98771158 0.39721054 3.25297437 2.21979063 0.13556017]]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


array([0])

**Matriz de Confusión**

Se procede con la evaluación o desempeño del modelo.

In [ ]:
ydp=mnb.predict(XD)  #Hazme el pronóstico de los positivos y negativos.
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

Ex=((VN+VP)/(VN+VP+FN+FP))*100
print("La exactitud del modelo es del:",Ex)

TE=((FN+FP)/(VN+VP+FN+FP))*100
print("La tasa de error del modelo es del;",TE)

Sen=(VP/(VP+FN))*100
print("La sensibilidad del modelo es del:",Sen)

Esp=(VN/(VN+FP))*100
print("La especificidad del modelo es del:",Esp)

Pre=(VP/(VP+FP))*100
print("La precisión del modelo es del:",Pre)

Pren=(VN/(VN+FN))*100
print("La precisión negativa del modelo es del:",Pren)

[[1365  215]
 [ 224  277]]
La exactitud del modelo es del: 78.90437289764536
La tasa de error del modelo es del; 21.095627102354637
La sensibilidad del modelo es del: 55.28942115768463
La especificidad del modelo es del: 86.39240506329115
La precisión del modelo es del: 56.30081300813008
La precisión negativa del modelo es del: 85.90308370044053


**Análisis de Resultados**
La base de datos analizada consta de 2.081 registros, donde el 76% corresponde a "0" y el 24% a "1", reflejando una distribución desbalanceada de clases.

Las variaciones en las principales categorías se presentan en los siguientes intervalos:

Día de la semana: 5.83, 5.94 y 1.93, 1.91.
ViolCat:10.97, 6.31 y 3.93, 1.98.
CrashType:6.09, 4.33 y 2.34, 2.21.
El modelo alcanzó una exactitud del 80%, con un 20% de error. Se destacan su especificidad 86% y precisión negativa 86%, lo que indica una alta capacidad para predecir correctamente la ausencia de accidentes en autopistas.

En una prueba con la combinación 5,1,1,3,4,0, el modelo predijo correctamente que no ocurriría un accidente. Los indicadores adicionales de desempeño están detallados en secciones previas.